### ISSUE TITLE
hipSYCL/sycl/buffer.hpp:185 "Writing back to external locations (not passed to " "buffer at construction) is unimplemented"' failed.

### ISSUE THREAD

**Issue by psath** [AdaptiveCpp/AdaptiveCpp#369]
This only happens sporadically (Out of 100 runs last night, 14 gave an assertion failure, 2 just segfaulted.) All of my kernels that either do an explicit `cgh.copy` afterwards seem to be fine. This is the only one that relies on any destructor-based copyback.

ROCm 3.8, Relatively new Clang/libomp from Github. PRs #303 #337 #362 merged in. `--hipsycl-target=omp;hip:gfx803,gfx900`

This unit test is writing back to a scalar variable to double-check the accessor offset.  (In the wider context of a work-partitioning framework). The code basically looks like this. (I can work to extract a MWE tomorrow, if needed.)

```
sycl::buffer<int, 1> A = ... //(passed in by shared_ptr. heap-allocated, associated with heap-allocated host pointer, both with lifetimes >> the test function's scope. This data gets in and out of the kernels fine with separate cgh.copy calls
int64_t length = ...//passed in by value
int64_t offset = ...//passed in by value

sycl::queue q;

int scalar = 0
{ //data scope for scalar_buf
  sycl:buffer<int> scalar_buf(&scalar, sycl::range<1>{1});
  q.submit([&](sycl::handler &cgh) {
    auto scalar_acc = scalar_acc.get_access<sycl::access::mode::discard_write>(cgh);
    auto A_acc = A.get_access<sycl::access::mode::read>(cgh, sycl::range<1>(length), sycl::id<1>(offset));
    cgh.single_task<class offset_check>([=](){
      scalar_acc[0] = A_acc[0];
    });
  });
} // <-- Crashes here, during buffer destruction. scalar_buf is the only one that should be destroyed here.
```

It breaks regardless of whether scalar is stack-allocated or heap- and regardless of whether i use a `single_task` or `parallel_for`

I've managed to catch an instance of the assertion failure in GDB. The `writeback_ptr` is the correct address of scalar.
`get_host_device()` returns
```
$8 = {_backend = {id = (hipsycl::rt::backend_id::hip | hipsycl::rt::backend_id::omp | unknown: 316),
    sw_platform = hipsycl::rt::api_platform::cuda, hw_platform = (unknown: 368)}, _device_id = 0}
```
But attempting to check the other half of the conditional triggers another assertion failure
```
p data->get_memory(get_host_device())
unit_tests: /path/to/hipSYCL/runtime/data.hpp:511: Memory_descriptor hipsycl::rt::data_region<void *>::get_memory(hipsycl::rt::device_id) const [Memory_descriptor = void *]: Assertion `has_allocation(dev)' failed.
```

Am I running into the race with the DAG flush that the comment above this assertion alludes to?

**Comment by illuhad** [comment]
> $8 = {_backend = {id = (hipsycl::rt::backend_id::hip | hipsycl::rt::backend_id::omp | unknown: 316),
    sw_platform = hipsycl::rt::api_platform::cuda, hw_platform = (unknown: 368)}, _device_id = 0}

The device id is corrupted - you can see that it has an unknown hardware platform. Because of this, it cannot find a matching allocation to write back to and crashes later.
Maybe try with address sanitizer to check what happens to that device id?

**Comment by psath** [comment]
Thanks for the suggestion, I'll see if I can pin down what's corrupting that state.

**Comment by illuhad** [comment]
> `scalar_acc[0] = A_acc[0];`

In SYCL 1.2.1 indexing is always done from the base of the allocation, regardless of whether you have an offset in your accessor. This is to be consistent when using accessors with offsets with kernel offsets. So, if your offset is not 0, you are reading invalid memory here (but this probably won't explaing the corrupted device id).

The behavior of indexing an accessor when an offset is present might change in future versions of SYCL.

**Comment by psath** [comment]
In practice the offset is zero when the device id is getting corrupted. Will be tracing back the corruption after some meetings today.

However that's an extremely useful correction to my mental model that I will fold back into our unit tests. Thanks!. I had thought putting the offset into the accessor was the equivalent to doing pointer math, i.e. accessor[idx] referred to buffer[offset+idx]. You're saying I have to still use accessor[offset+idx] (at least in the 1.X major revisions)

**Comment by psath** [comment]
I'm still poking at this as I have time. I've worked around the assertion failure by explicitly doing a copy command group. However, my debug build of the hipSYCL runtime library reliably throws out a buffer_memory_requirement with a garbage access_range. I've never managed to catch one being constructed improperly, so it still seems like something's trashing the stack after the fact.

Haven't gotten anything concrete out of GDB or valgrind/helgrind, and a clang `-fsanitize=address` build of the hipSYCL runtime is hanging after/during `[hipSYCL Info] backend_manager: Registering HIP backend...` in an `munmap` loop. Will get back to it after US Thanksgiving break.

**Comment by illuhad** [comment]
The address sanitizer may have trouble with GPU builds as in that case you have external devices randomly writing to memory and lots of things going on at kernel driver level that it doesn't know about. Does it also reproduce with a CPU-only build?

> You're saying I have to still use accessor[offset+idx] (at least in the 1.X major revisions)

Yes. The reason is that if you launch a kernel with an offset, you want to be able to do:
```
auto accessor = buff.get_access<...>(cgh, myrange, myoffset);
parallel_for(myrange, myoffset, [=](sycl::id idx) {
  accessor[idx] = 0;
});
```
This only works if the accessor doesn't add the offset, since the offset will already be added to the index passed into the kernel.

**Comment by zjin-lcf** [comment]
Is "psath" Paul Sathre ?  I'd like to say Hi to Paul.

**Comment by psath** [comment]
That's... extremely disconcerting and off-topic. Have we met? While I'm not exactly trying to anonymize my profile (it has papers associated with it after all), this is not an appropriate place for a personal discussion. Send me an email.

@illuhad can you expunge the PII from the thread? I'd appreciate it.

**Comment by illuhad** [comment]
@psath done.

**Comment by psath** [comment]
All signs point to it being some issue with host-associated buffers, as long as I don't initialize the buffer with a host pointer (neither stack nor heap), and instead just use a copy command group to transfer back after the kernel, I don't run into assertion failures or segfaults.

If I can knock out our bigger tests using this workaround before the next US holiday shutdown, I'll loop back and keep looking for the misbehaving thread(s) in my spare cycles.

**Comment by illuhad** [comment]
Thanks for the update! If there are no host pointers, there will be no writeback and the destruction of the buffer will be much simpler.

I've just had a stack trace from the nbody-sycl application from #380 (which crashes in a similar manner at writeback at buffer destruction) where it segfaults inside `hipMemcpyAsync`. And then my AMD GPU, which also powers my display, started hanging and I had to reboot ;) So, it might indeed be the case that something is wrong with the data transfer.

**Comment by psath** [comment]
> The address sanitizer may have trouble with GPU builds as in that case you have external devices randomly writing to memory and lots of things going on at kernel driver level that it doesn't know about. Does it also reproduce with a CPU-only build?
> 
> > You're saying I have to still use accessor[offset+idx] (at least in the 1.X major revisions)
> 
> Yes. The reason is that if you launch a kernel with an offset, you want to be able to do:
> 
> ```
> auto accessor = buff.get_access<...>(cgh, myrange, myoffset);
> parallel_for(myrange, myoffset, [=](sycl::id idx) {
>   accessor[idx] = 0;
> });
> ```
> 
> This only works if the accessor doesn't add the offset, since the offset will already be added to the index passed into the kernel.

Ok, this looks like what I was trying to do in the first place, where the index is relative to the offset buffer. I just have to apply it to the kernel lambda in addition to the accessor. (Effectively `myoffset` is just added to the initial value of `idx` -- by the `parallel_for` *not the accessor decl/def* -- so it would apply anywhere I used that index, right?)

**Comment by illuhad** [comment]
>  Ok, this looks like what I was trying to do in the first place, where the index is relative to the offset buffer. I just have to apply it to the kernel lambda in addition to the accessor. (Effectively myoffset is just added to the initial value of idx -- by the parallel_for not the accessor decl/def -- so it would apply anywhere I used that index, right?)

Yes, that's precisely what happens. Note that with SYCL 2020, offsets to kernels may... change. Unfortunately I cannot comment publicly on the details yet.

**Comment by psath** [comment]
> Thanks for the update! If there are no host pointers, there will be no writeback and the destruction of the buffer will be much simpler.
> 
> I've just had a stack trace from the nbody-sycl application from #380 (which crashes in a similar manner at writeback at buffer destruction) where it segfaults inside `hipMemcpyAsync`. And then my AMD GPU, which also powers my display, started hanging and I had to reboot ;) So, it might indeed be the case that something is wrong with the data transfer.

I'd be interested to see what, if anything, differs between the DAGs of 1) Kernel with writes to host-associated buffer, that then destructs, vs. 2) Kernel with writes to a non-host-associated buffer, that then explicitly copies to the host, that then destructs.

If the buffer's accessor is `discard_write` they should have similar incoming requirements, and both require that a) the kernel run and b) any changes to the buffer be written back to a specified host pointer, before c) the buffer is destructed. However, the first is done as a single queue submission, the second is done as two (one for kernel, one for explicit copy).

**Comment by illuhad** [comment]
Good question.

> If the buffer's accessor is discard_write they should have similar incoming requirements

Agreed.

>  a) the kernel run and b) any changes to the buffer be written back to a specified host pointer, before c) the buffer is destructed

Yes

>  However, the first is done as a single queue submission, the second is done as two (one for kernel, one for explicit copy).

Well.. the writeback also performs a submission (and then immediately waits for it to complete, to obtain a synchronous behavior) inside the destructor, so you always have two submission from the point-of-view of the hipSYCL runtime.

The most substantial difference is that they go through two quite different mechanisms:
* for the explicit copy, hipSYCL assumes the stance that if the user explicitly wants a copy, it will always be executed exactly as requested since it was explicitly formulated. So, basically the copy instruction then goes straight to the backend which then executes it.
* with writeback, the semantics are more similar to `handler::update_host()`, i.e. the writeback is treated as an implicit requirement: Before execution can continue, make sure that the requested data is present on the host, and do whatever is necessary for that. This means the runtime can make various optimizations, such as only transferring a smaller region of the buffer if parts are already up-to-date on the host, or even optimizing the copy away entirely. Basically the same thing happens as when you construct an accessor in a kernel: the SYCL implementation translates an implicit requirement into 0 or more actual operations.

I wonder if it causes an issue if we submit operations for the buffer when the buffer is already in the destructor. But I have just checked, the actual `data_region` object, which is the only thing the runtime really cares about, is still alive at that point, and it's a `shared_ptr`, so giving the runtime another copy during submission should make everything safe.

Do you see a difference between optimization levels? The `nbody-sycl` error doesn't trigger with disabled optimizations.

**Comment by psath** [comment]
> I wonder if it causes an issue if we submit operations for the buffer when the buffer is already in the destructor. But I have just checked, the actual `data_region` object, which is the only thing the runtime really cares about, is still alive at that point, and it's a `shared_ptr`, so giving the runtime another copy during submission should make everything safe.
> 
> Do you see a difference between optimization levels? The `nbody-sycl` error doesn't trigger with disabled optimizations.

Yeah IIRC every time I've had it in GDB, the `data_region` is fine, it's the `buffer_memory_requirement` having its `access_range` tampered to something like `{1, 1, 2478978563054}` that breaks everything.

I admit, this test is deep inside a bigger framework configured with cmake that I'm building with `-DCMAKE_BUILD_TYPE=Debug`, but I haven't actually validated that it's doing the SYCL portions with `-O0` rather than 2 or 3. I've also swapped out Clang 8,9,10,11, and 12-indev, and GCC 8.3.0 as the host compiler (typically hipSYCL has been compiled with GCC 8.3, until the more recent tests when I tried to add AddressSanitizer with 12-indev), so that's definitely a variable to isolate.

When I get back to it I'll have to just create an isolated MWE that I can test different options of compiler/flags on more cleanly.

**Comment by illuhad** [comment]
Okay, so I managed to reproduce this on CUDA as well.

However, I discovered the following workaround which worked for me, both on CUDA and ROCm: Insert an additional `queue::wait()` right after the buffer destructor (and its writeback) executes, and before the queue goes out of scope.
Can you check if this also works for your case?

EDIT: A more convenient workaround that also works for me is to replace the code inside `if(destructor_waits){ ... }` in `~buffer_impl()` in `buffer.hpp` with the following:
```
rt::application::dag().flush_sync();
rt::application::dag().wait();
```
This does a full sync of the DAG instead of only waiting for the DAG nodes that are referring to this buffer.

**Comment by illuhad** [comment]
Can you please check if the issue persists with PR #395?

**Comment by psath** [comment]
> Can you please check if the issue persists with PR #395?

Rebuilding and running now.

When I tested with the attached MWE last night (pre #395), without optimization flags
[c-vec-add.cpp.txt](https://github.com/illuhad/hipSYCL/files/5630759/c-vec-add.cpp.txt)



with `-DUSE_HOST_PTR`: 381/1000 failures (19 segfaults, the rest `free(): invalid pointer` faults)
with `-DUSE_HOST_PTR -DFORCE_FLUSH`: 86/1000 failures (7 segfaults, the rest `free(): invalid pointer` faults)
with `-DFORCE_FLUSH`: 0/1000 failures
no defines (explicit copy, no flush): 0/1000 failures

And when it didn't fail, the hostptr version with the explicit flush had never copied the results back to the host, but all of the explicit copies (both explicitly and implicitly flushed) moved things back correctly. Interestingly the hostptr version *without* any explicit flush seemed to get the data back (when it didn't fail).

So it definitely had something to do with the host pointer, and flushing *helped* but didn't resolve it entirely. Looks like a use-after-free and/or double-free issue.

**Comment by psath** [comment]
Yep #395 resolves it in the MWE. The USE_HOST_PTR version with and without the flush passed all 1000 tests (and reliably copied the data back to the host)

Will try it in the full framework and mark as resolved if that also succeeds.

**Comment by illuhad** [comment]
Nice!! Thank you! That was a lot of time that went into cracking this one ;)

**Comment by psath** [comment]
Hmmm, I do still see if very rarely (2/100) in my full test (where the writeback is only a scalar, and the data buffer creation and destruction is far-removed from the kernel). I'll try to snag it in GDB

Shoot, and I had the hipsycl debug output turned off. Retrying with those diagnostics on.

There's a chance my framework is doing something to bash memory as well.

```C
  int meh = 0;
  //meh has to be wrapped in a buffer so we can give it an accessor, and that buffer needs to be scoped so we can force it to finish being read back
  {
    cl::sycl::buffer<int> meh_buf(&meh, cl::sycl::range<1>{1});
//    cl::sycl::buffer<int> meh_buf(cl::sycl::range<1>{1});
    //capture by reference so we can write to meh and don't create unnecessary copies
    a_queue.submit([&](cl::sycl::handler &cgh) {
      auto meh_acc = meh_buf.get_access<cl::sycl::access::mode::discard_write>(cgh);
      auto v1_acc = v1.get_access<cl::sycl::access::mode::read>(cgh, cl::sycl::range<1>(1), cl::sycl::id<1>(v1_offset));
      //cgh.single_task<class offset_check>([=](){
      cgh.parallel_for<class offset_check>(cl::sycl::range<1>(1), cl::sycl::id<1>(v1_offset), [=](cl::sycl::id<1> i){
        meh_acc[0] = v1_acc[0];
      });
    });
//    a_queue.submit([&](cl::sycl::handler &cgh) {
//      auto meh_acc = meh_buf.get_access<cl::sycl::access::mode::read>(cgh);
//      cgh.copy(meh_acc, &meh);
//    });
//    a_queue.wait();
  }
```

**Comment by psath** [comment]
```shell
^[[;32m[hipSYCL Info] ^[[0mdata_region: constructed with page table dimensions 1 1 1
^[[;32m[hipSYCL Info] ^[[0mdag_manager: Checking DAG flush opportunity...
^[[;32m[hipSYCL Info] ^[[0mdag_manager: Submitting asynchronous flush...
^[[;32m[hipSYCL Info] ^[[0mdag_manager [async]: Flushing!
^[[;32m[hipSYCL Info] ^[[0mbuffer_impl::~buffer_impl: Preparing submission of writeback...
^[[;32m[hipSYCL Info] ^[[0mdag_builder: DAG contains operation: kernel: ZZ15vector_add_syclENKUlRN7hipsycl4sycl7handlerEE0_clES2_EUlNS0_2idILi1EEEE_
   MEM_REQ: Discard W global_buffer {0, 0, 0}+{1, 1, 1} #4
   MEM_REQ: RW global_buffer {0, 0, 0}+{1, 1, 1000} #4
   MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 1000} #4
^[[;32m[hipSYCL Info] ^[[0m    --> requires: MEM_REQ: Discard W global_buffer {0, 0, 0}+{1, 1, 1} #4
^[[;32m[hipSYCL Info] ^[[0m    --> requires: MEM_REQ: RW global_buffer {0, 0, 0}+{1, 1, 1000} #4
^[[;32m[hipSYCL Info] ^[[0m    --> requires: MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 1000} #4
^[[;32m[hipSYCL Info] ^[[0mdag_builder: DAG contains operation: MEM_REQ: Discard W global_buffer {0, 0, 0}+{1, 1, 1} #4
^[[;32m[hipSYCL Info] ^[[0mdag_builder: DAG contains operation: MEM_REQ: RW global_buffer {0, 0, 0}+{1, 1, 1000} #4
^[[;32m[hipSYCL Info] ^[[0mdag_builder: DAG contains operation: MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 1000} #4
^[[;32m[hipSYCL Info] ^[[0mdag_manager [async]: Releasing dead users of data region 0x7fa85d2dc5c0
^[[;32m[hipSYCL Info] ^[[0mdag_manager [async]: Releasing dead users of data region 0x7fa814001380
^[[;32m[hipSYCL Info] ^[[0mdag_manager [async]: Releasing dead users of data region 0x7fa85d2dc450
^[[;32m[hipSYCL Info] ^[[0mdag_manager [async]: Submitting node to direct scheduler!
^[[;32m[hipSYCL Info] ^[[0mdag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: Discard W global_buffer {0, 0, 0}+{1, 1, 1} #4
unit_tests: /path/to/hipsycl/install/bin/../include/CL/../hipSYCL/sycl/buffer.hpp:185: hipsycl::sycl::detail::buffer_impl::~buffer_impl(): Assertion `false && "Writing back to external locations (not passed to " "buffer at construction) is unimplemented"' failed.
^[[;32m[hipSYCL Info] ^[[0mdag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: RW global_buffer {0, 0, 0}+{1, 1, 1000} #4
^[[;32m[hipSYCL Info] ^[[0mdag_direct_scheduler: Preparing deferred pointer of requirement node MEM_REQ: R global_buffer {0, 0, 0}+{1, 1, 1000} #4
```

This is from a later kernel that does a vectorAdd with another scalar writeback for sanity checking.
```C
  {
    cl::sycl::buffer<int> meh_buf(&meh, cl::sycl::range<1>{1});
//    cl::sycl::buffer<int> meh_buf(cl::sycl::range<1>{1});
    a_queue.submit([&](cl::sycl::handler &cgh) {
      auto meh_acc = meh_buf.get_access<cl::sycl::access::mode::discard_write>(cgh);
      auto v1_acc = v1.get_access<cl::sycl::access::mode::read_write>(cgh, cl::sycl::range<1>(v1_length), cl::sycl::id<1>(v1_offset));
      auto v2_acc = v2.get_access<cl::sycl::access::mode::read>(cgh, cl::sycl::range<1>(v2_length), cl::sycl::id<1>(v2_offset));
      //The parallel_for needs a name and the number of elements to process, and then a name for the id variable
      cgh.parallel_for<class vector_add>(cl::sycl::range<1>(v1_length), cl::sycl::id<1>(v1_offset), [=](cl::sycl::id<1> i) {
        v1_acc[i] += v2_acc[i];
        meh_acc[0] = v1_acc[i];
      });
    });
//    a_queue.submit([&](cl::sycl::handler &cgh) {
//      auto meh_acc = meh_buf.get_access<cl::sycl::access::mode::read>(cgh);
//      cgh.copy(meh_acc, &meh);
//    });
//    a_queue.wait();
  }
```

**Comment by psath** [comment]
> > $8 = {_backend = {id = (hipsycl::rt::backend_id::hip | hipsycl::rt::backend_id::omp | unknown: 316),
> > sw_platform = hipsycl::rt::api_platform::cuda, hw_platform = (unknown: 368)}, _device_id = 0}
> 
> The device id is corrupted - you can see that it has an unknown hardware platform. Because of this, it cannot find a matching allocation to write back to and crashes later.
> Maybe try with address sanitizer to check what happens to that device id?

Seems to be the same thing happening still. The writeback pointer in data is correct but the returned host device id is mangled
`{_backend = {id = (hipsycl::rt::backend_id::hip | hipsycl::rt::backend_id::omp | unknown: 316), hw_platform = hipsycl::rt::hardware_platform::rocm, sw_platform = (unknown: 368)}, _device_id = 0}`
Shouldn't `hipsycl::sycl::detail::get_host_device()` strictly return `rt::hardware_platform::cpu, rt::api_platform::omp)`?

**Comment by psath** [comment]
Adding a scalar writeback to the MWE and battering it again, to see if I can reproduce outside our framework.

Edit: Made it through 512/1000 hostptr+flush tests just fine before something completely hung my machine, necessitating a cold IPMI reboot /grimace

**Comment by illuhad** [comment]
> Shouldn't hipsycl::sycl::detail::get_host_device() strictly return rt::hardware_platform::cpu, rt::api_platform::omp)?

It does. See https://github.com/illuhad/hipSYCL/blob/347f1e1b05050aa5440a6b657ad28592a3dbc496/include/hipSYCL/sycl/device.hpp#L56
Anything else means something is going terribly wrong.

**Comment by psath** [comment]
> > Shouldn't hipsycl::sycl::detail::get_host_device() strictly return rt::hardware_platform::cpu, rt::api_platform::omp)?
> 
> It does. See
> 
> https://github.com/illuhad/hipSYCL/blob/347f1e1b05050aa5440a6b657ad28592a3dbc496/include/hipSYCL/sycl/device.hpp#L56
> 
> 
> Anything else means something is going terribly wrong.

Yes, that's what concerns me.

I'm not getting any hits on the isolated version. Our big code must be doing something untoward that's precipitating out here. There's a lot of passing shared_ptrs to SYCL types through a void*-casted C-like interface going on. So a fair chance that something is getting reinterpret_casted in a busted way, perhaps in the explicit copies before the kernel.

I've also just turned up an issue writing *anything* to my second GPU (R9 nano / gfx803), which may be related. Until now I've just been using the Vega 10 (gfx900) that SYCL defaults to, but we're intending to co-schedule to both. Once I get some new insights I'll report back.

**Comment by illuhad** [comment]
> I'm not getting any hits on the isolated version. Our big code must be doing something untoward that's precipitating out here. There's a lot of passing shared_ptrs to SYCL types through a void*-casted C-like interface going on. So a fair chance that something is getting reinterpret_casted in a busted way, perhaps in the explicit copies before the kernel.

hm.. at least it seems that PR #395 is right in itself as it seems to make writeback more robust, even if there might be other issues.

> I've also just turned up an issue writing anything to my second GPU (R9 nano / gfx803), which may be related. Until now I've just been using the Vega 10 (gfx900) that SYCL defaults to, but we're intending to co-schedule to both. Once I get some new insights I'll report back.

Ah, multi-GPU is something that hipSYCL's new runtime was designed for (the old one ... was not) but that has only been tested very lightly until now, and not routinely since my development machines aren't multi-GPU and it's also more complex to have such a setup in CI. So, it might be that there are some issues there. I can verify tomorrow whether a simple test code works with current `develop`.

**Comment by psath** [comment]
> > I'm not getting any hits on the isolated version. Our big code must be doing something untoward that's precipitating out here. There's a lot of passing shared_ptrs to SYCL types through a void*-casted C-like interface going on. So a fair chance that something is getting reinterpret_casted in a busted way, perhaps in the explicit copies before the kernel.
> 
> hm.. at least it seems that PR #395 is right in itself as it seems to make writeback more robust, even if there might be other issues.

Yes, it definitely eliminated a number of spurious issues! I ran a short (40ish) test of the full framework with a queue wait after each kernel, and didn't run into any writeback issues. But that's too short to be certain, will start a long one when I'm clocking out for the day. I've restructured the simple vector add I attached earlier to more directly match what we're doing (i.e. the data buffers are always explicitly transferred, and their lifetime far exceeds that of the scalar), so giving that a spin now.

> 
> > I've also just turned up an issue writing anything to my second GPU (R9 nano / gfx803), which may be related. Until now I've just been using the Vega 10 (gfx900) that SYCL defaults to, but we're intending to co-schedule to both. Once I get some new insights I'll report back.
> 
> Ah, multi-GPU is something that hipSYCL's new runtime was designed for (the old one ... was not) but that's only tested very lightly until now, and not routinely since my development machines aren't multi-GPU and it's also more complex to have such a setup in CI. So, it might be that there are some issues there. I can verify tomorrow whether a simple test code works with current `develop`.

Thanks, I'll be curious to see if it works for you. We are very invested in multi-AMD (and later, hopefully mixed AMD+Nvidia if that can be supported)

I'm going to point the vectorAdd at just the R9 after another pass on the Vega and see if it misbehaves the same way it does in the full framework. (FWIW the specific fault I'm seeing on the R9 is a write into mapped `/dev/dri/renderD129` by a worker thread inside `libamdhip64.so`, that *does not* turn up if I use the device directly via the HIP MatrixTranspose sample.) If I can't pin it to something I'm doing wrong I can open a separate issue for it.

**Comment by psath** [comment]
> > > I'm not getting any hits on the isolated version. Our big code must be doing something untoward that's precipitating out here. There's a lot of passing shared_ptrs to SYCL types through a void*-casted C-like interface going on. So a fair chance that something is getting reinterpret_casted in a busted way, perhaps in the explicit copies before the kernel.
> > 
> > 
> > hm.. at least it seems that PR #395 is right in itself as it seems to make writeback more robust, even if there might be other issues.
> 
> Yes, it definitely eliminated a number of spurious issues! I ran a short (40ish) test of the full framework with a queue wait after each kernel, and didn't run into any writeback issues. But that's too short to be certain, will start a long one when I'm clocking out for the day. I've restructured the simple vector add I attached earlier to more directly match what we're doing (i.e. the data buffers are always explicitly transferred, and their lifetime far exceeds that of the scalar), so giving that a spin now.

Ok, new MWE is 
[c-vec-add-w-scalar.cpp.txt](https://github.com/illuhad/hipSYCL/files/5632138/c-vec-add-w-scalar.cpp.txt)

**1** out of 1000 USE_HOST_PTR tests without a flush threw the issue. Full log attached. So it's exceedingly rare now, but unless I'm doing something wrong in the MWE, not exclusive to our larger framework.
[hostptr-noflush.0.err.txt](https://github.com/illuhad/hipSYCL/files/5632141/hostptr-noflush.0.err.txt)

**Comment by illuhad** [comment]
Oh dear.. it will be very difficult to debug something that happens so rarely. So far I could not reproduce. It might be a good idea to have a device emulation backend that will run on CPU, since sanitizers don't work well with GPU backends as there is too much going on that they don't know about.

> and later, hopefully mixed AMD+Nvidia if that can be supported

The runtime is designed to support multiple backends simultaneously (which already happens if you target both omp and hip). The problem is compiling kernels for multiple device backends. hipSYCL relies on clang's CUDA/HIP driver with an additional clang plugin to add support for SYCL constructs, but in the end it just executes `clang -x cuda` or `clang -x hip`. And clang does not support compiling both for CUDA and HIP simultaneously since they are two different clang drivers.
We have discussed this previously, and it might be possible to circumvent this limitation, but it would boil down to reinventing parts of the toolchain that clang already has, such as embedding compiled device kernels in the host application. This would require a significant engineering effort. On the other hand, in practice it seems a rather uncommon use case to have both NVIDIA and AMD GPUs in the same node (at least in HPC). Because of this I don't currently see this as a good investment of the time it would consume to get this working.
EDIT: If you can get your compiled kernels from some external source it might work, as hipSYCL supports CUDA/HIP interop

**Comment by illuhad** [comment]
Removing this from hipSYCL 0.9 milestone so that the release is not blocked by issues that are hard to reproduce like the remaining problem here.

**Comment by illuhad** [comment]
Still cannot reproduce in 2000 runs. There's a race condition in your MWE since all work items write to `meh_acc` at the same time. I don't know how that could corrupt host memory, but technically it's UB, and I don't know how AMD chips/ROCm react to that.
